### PR TITLE
Respect startup-file flag in Pkg.build and Pkg.test

### LINKED ISSUE #21823: Pkg.update() does not respect --startup-file=no
When transitioning from Julia v0.5 to v0.6, I find it useful to start julia v0.6 with the `--startup-file=no` flag. This works alright until I do for instance `Pkg.update()`, in the end of which julia tries to load the startup file, which in my case causes an error to be thrown. My solution is to comment out everything in the file, which is a bit awkward.

### PR DESCRIPTION
**Pull request by omus**
Fixes #21823 since `Pkg.update` calls `Pkg.build`

### CODE REVIEW

**Comment by tkelman** [comment]
probably worth adding a startup_file kwarg so you could turn this back on if you wanted (defaulting to off is better though)

**Review by yuyichao — DISMISSED** [review]
This does not fix https://github.com/JuliaLang/julia/issues/21823 since it still doesn't "respect --startup-file=no" .

**Comment by omus** [comment]
I'll add @tkelman's suggestion to support loading the startup file through a keyword.

**Comment by omus** [comment]
After thinking about it more I think think we shouldn't support a `startup_file` keyword. I don't think either `Pkg.test` or `Pkg.build` should depend on anything within the startup file since this would be a requirement which is not properly tracked by the package. Additionally, having the `startup_file` keyword exposes that both `Pkg.test` and `Pkg.build` are launching new Julia processes which may not be true in the future.

**Comment by omus** [comment]
> This does not fix #21823 since it still doesn't "respect --startup-file=no" .

Although the title of the issue wants `Pkg.update` to respect the startup-file flag all the OP is asking for is that his startup-file is not loaded during `Pkg.update`. I don't think there is a good use case for loading the startup-file here.

**Comment by tkelman** [comment]
I'm thinking about deployments in unusual environments that may be using juliarc to set up custom library paths, that kind of thing. I wouldn't think it would be a commonly used flag, but I don't see an issue with exposing that separate processes are getting launched. If that changes we can deprecate the keyword arguments - we already have one for code coverage which is a similar situation.

**Comment by omus** [comment]
> I'm thinking about deployments in unusual environments that may be using juliarc to set up custom library paths, that kind of thing.

A good point. Since several Pkg functions eventually call `Pkg.build` (like `Pkg.update`) we would also need to have the keyword available for those functions as well.

It seems that the easiest solution would be to just respect the `--startup-file` flag from the parent Julia process. Unfortunately this means the default will be to load the startup-file but at least users will have the option.

**Comment by omus** [comment]
The `JLOptions().startupfile` uses 0 for unset, 1 for "yes", and 2 for "off.

**Comment by tkelman** [comment]
This version seems like an improvement over what we've got now at least (and would make a little more sense to consider backporting even, as opposed to new keyword args). Could use a test though.

**Comment by omus** [comment]
> Could use a test though.

Is it safe to temporarily modify `abspath(JULIA_HOME,"..","etc","julia","juliarc.jl")` for a test?

**Comment by tkelman** [comment]
could set HOME in a `withenv` ?

File: base/pkg/entry.jl [hunk 1]
```diff
@@ -719,7 +720,7 @@ function test!(pkg::AbstractString,
                     --compilecache=$(Bool(Base.JLOptions().use_compilecache) ? "yes" : "no")
                     --check-bounds=yes
                     --startup-file=$(Base.JLOptions().startupfile != 2 ? "yes" : "no")
-                    $test_path
+                    --eval $code
```
Comment (tkelman) [inline base/pkg/entry.jl]: why is this change necessary?

Comment (omus) [inline base/pkg/entry.jl]: Not strictly necessary but does make the "runtests.jl" file run inside of a module which means variables defined within the startup-file can only be accessed via `Main.*`. I made this modification to make it harder to accidentally reference a variable within the tests which was defined or imported within the startup-file.

**Comment by tkelman** [comment]
the appveyor timeouts look like redirection to a file was necessary

**Comment by omus** [comment]
> could set HOME in a `withenv`?

Looks like for Windows you need to set [`USERPROFILE`](http://docs.libuv.org/en/v1.x/misc.html#c.uv_os_homedir).

**Comment by omus** [comment]
I'll try to sort out the issues tomorrow

**Comment by omus** [comment]
Resolved the following issues:
- Needed to use `escape_string` in `Pkg.test` to work correctly with Windows paths
- `isdefined(::Symbol)` was deprecated which caused my tests to fail on Linux and Mac due to starting julia with `--depwarn=error`

File: base/pkg/entry.jl (hunk 1 above)
Comment (tkelman) [inline base/pkg/entry.jl]: that did necessitate the messy escape_string business though

**Comment by omus** [comment]
Ready to merge

File: base/pkg/entry.jl (hunk 1 above)
Comment (tkelman) [inline base/pkg/entry.jl]: is running the runtests.jl inside a module via evalfile going to result in any packages' tests behaving differently?

**Comment by omus** [comment]
I'll merge this today unless there are objections. I need to close one of the 1800 issues ;)

**Comment by tkelman** [comment]
had a now-collapsed question on the change to using evalfile

> is running the runtests.jl inside a module via evalfile going to result in any packages' tests behaving differently?

I guess we'll find out, but I'm a bit concerned if packages will notice the change.

**Comment by omus** [comment]
> is running the runtests.jl inside a module via evalfile going to result in any packages' tests behaving differently?

Packages which assume the tests are running in the `Main` module could break. For example Compat.jl will run require two changes in runtests.jl lines [1570](https://github.com/JuliaLang/Compat.jl/blob/master/test/runtests.jl#L1570) and [1855](https://github.com/JuliaLang/Compat.jl/blob/master/test/runtests.jl#L1855). I'll make a Compat.jl PR.

**Comment by omus** [comment]
I'll make a NEWS entry to make debugging this change easier.

**Comment by tkelman** [comment]
In that case I don't personally think this behavior change to Pkg.test is worth the disruption - CI will virtually never have a juliarc, after all.

**Comment by iamed2** [comment]
@tkelman Are you referring to just the `evalfile` part?

**Comment by iamed2** [comment]
Tests really shouldn't be assuming `Main`, especially since `current_module()`/`@__MODULE__` is available.

**Comment by tkelman** [comment]
> Are you referring to just the `evalfile` part?

Right, the change in which module things run in. Package tests have been run in this way for as long as Pkg.test has existed, and tweaking them doesn't seem to be accomplishing much of anything aside from this minor point about juliarc isolation.

**Comment by omus** [comment]
I'd like to get the juliarc isolation included but I can accept that it is a controversial change. I'll remove that change from this PR so the remainder of the changes can at least be merged.

**Comment by tkelman** [comment]
Sorry, yes, propagating `--startup-file` is a definite improvement, but is slightly independent of whether or not `Pkg.test` runs code via `evalfile`

**Comment by tkelman** [comment]
LGTM now, though the NEWS bit should move to where the evalfile change goes

**Comment by omus** [comment]
> LGTM now, though the NEWS bit should move to where the evalfile change goes
Thanks. I just noticed that myself.

**Review by tkelman — APPROVED** [review]
+1 to merge if CI is green
